platform/ios/src/MGLMapView.mm

```diff
@@ -5375,33 +5374,50 @@ void onCameraDidChange(mbgl::MapObserver::CameraChangeMode mode) override {
     }
 
     void onWillStartLoadingMap() override {
-        [nativeView willStartLoadingMap];
+        [nativeView mapViewWillStartLoadingMap];
     }
 
     void onDidFinishLoadingMap() override {
-        [nativeView didFinishLoadingMap];
+        [nativeView mapViewDidFinishLoadingMap];
     }
 
-    void onDidFailLoadingMap() override {
-        [nativeView didFailLoadingMap];
+    void onDidFailLoadingMap(mbgl::MapObserver::ErrorType type, const std::string& message) override {
+        NSError *error;
+        NSString *errorDescription = @(message.c_str());
+        MGLErrorCode errorCode;
+        switch (type) {
+        case mbgl::MapObserver::ErrorType::ParseStyle:
+            errorCode = MGLErrorCodeParseStyleFailed;
+            break;
+        case mbgl::MapObserver::ErrorType::NotFound:
+            errorCode = MGLErrorCodeNotFound;
+            break;
+        case mbgl::MapObserver::ErrorType::LoadingStyle:
+            errorCode = MGLErrorCodeLoadStyleFailed;
+            break;
+        }
+        error = [NSError errorWithDomain:MGLErrorDomain code:errorCode userInfo:errorDescription ? @{
+            NSLocalizedDescriptionKey: errorDescription,
```

Comment:
Thank you so much for this! 👍 It turns out the [error user info dictionary](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/ErrorHandlingCocoa/ErrorObjectsDomains/ErrorObjectsDomains.html#//apple_ref/doc/uid/TP40001806-CH202-SW3) is called that because its contents are often displayed to the user with a localized description and sometimes a recovery suggestion. Here are some suggestions for improvements, taking advantage of the error type and assuming `message` comes directly from an `std::exception_ptr`, as suggested in https://github.com/mapbox/mapbox-gl-native/pull/8418#issuecomment-286820676:

* `ParseStyle`:
  * `NSLocalizedDescriptionKey`: “The map failed to load because the style is corrupted.”
  * `NSLocalizedFailureReasonErrorKey`: `@(message.c_str())`
* `NotFound`:
  * `NSLocalizedDescriptionKey`: “The map failed to load because the style can’t be found or is incompatible.”
  * `NSLocalizedFailureReasonErrorKey`: `@(message.c_str())`
* `LoadingStyle`:
  * `NSLocalizedDescriptionKey`: `@(message.c_str())`

For the localized descriptions, we can use the `NSLocalizedStringWithDefaultValue()` macro (as described [here](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/ios/DEVELOPING.md#adding-user-facing-text)) to potentially localize the error message.

Let me know if you’d prefer that I make these changes myself.